src/Definition.cpp

```diff
@@ -0,0 +1,210 @@
+        for (size_t i = 0; i < args.size(); ++i) {
+            args[i] = mutator->mutate(args[i]);
+        }
+
+        schedule.mutate(mutator);
+
+        if (domain.defined()) {
+            domain.mutate(mutator);
+        }
+
+        for (Specialization &s : specializations) {
+            if (s.condition.defined()) {
+                s.condition = mutator->mutate(s.condition);
+            }
+            s.definition.mutate(mutator);
+        }
+    }
+};
+
+template<>
+EXPORT RefCount &ref_count<DefinitionContents>(const DefinitionContents *d) {
+    return d->ref_count;
+}
+
+template<>
+EXPORT void destroy<DefinitionContents>(const DefinitionContents *d) {
+    delete d;
+}
+
+Definition::Definition() : contents(new DefinitionContents) {}
+
+Definition::Definition(const IntrusivePtr<DefinitionContents> &ptr) : contents(ptr) {
+    internal_assert(ptr.defined())
+        << "Can't construct Function from undefined DefinitionContents ptr\n";
+}
+
+Definition::Definition(const std::vector<Expr> &args, const std::vector<Expr> &values,
+                       const ReductionDomain &rdom, bool is_init) : contents(new DefinitionContents) {
+    contents->is_init = is_init;
+    contents->values = values;
+    contents->args = args;
+    contents->domain = rdom;
+}
+
+// Return deep-copy of Definition
+Definition Definition::deep_copy(
+        std::map<IntrusivePtr<FunctionContents>, IntrusivePtr<FunctionContents>> &copied_map) const {
+    internal_assert(contents.defined()) << "Cannot deep-copy undefined Definition\n";
+
+    Definition copy;
+    copy.contents->is_init = contents->is_init;
+    copy.contents->values = contents->values;
+    copy.contents->args = contents->args;
+    copy.contents->schedule = contents->schedule.deep_copy(copied_map);
+
+    // Definition's domain is the same as the one pointed by its schedule.
+    internal_assert((contents->is_init == !contents->domain.defined()) || !contents->is_init)
+        << "Init definition should not have a reduction domain\n";
+
+    internal_assert((!contents->domain.defined() && !contents->schedule.reduction_domain().defined()) ||
+                    (contents->domain.defined() && contents->schedule.reduction_domain().same_as(contents->domain)))
+        << "Definition should point to the same reduction domain as its schedule\n";
+    // We don't need to deep-copy the reduction domain since we've already done
+    // it when deep-copying the schedule above
+    copy.contents->domain = copy.schedule().reduction_domain();
+
+    // Deep-copy specializations
+    for (const auto &s : contents->specializations) {
+        Specialization s_copy;
+        s_copy.condition = s.condition;
+        s_copy.definition = s.definition.deep_copy(copied_map);
+        copy.contents->specializations.push_back(std::move(s_copy));
+    }
+    return copy;
+}
+
+bool Definition::is_init() const {
+    internal_assert(!contents->is_init || (contents->is_init && !contents->domain.defined()))
```

Comment:
I think contents->is_init should be assumed to be true in the second part of the check, similar to above.